File: Python/Problem_107.py
class DSU:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, x: int) -> int:
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    def union(self, a: int, b: int) -> bool:
        ra, rb = self.find(a), self.find(b)
        if ra == rb:
            return False
        if self.rank[ra] < self.rank[rb]:
            ra, rb = rb, ra
        self.parent[rb] = ra
        if self.rank[ra] == self.rank[rb]:
            self.rank[ra] += 1
        return True

def solve(path: str = "network.txt") -> int:
    rows = []
    with open(path, "r", encoding="utf-8") as f:
        for line in f:
            rows.append(line.strip().split(","))

    n = len(rows)
    edges = []
    total = 0

    for i in range(n):
        for j in range(i + 1, n):
            val = rows[i][j]
            if val != "-":
                w = int(val)
                total += w
                edges.append((w, i, j))

    edges.sort()
    dsu = DSU(n)

    mst = 0
    used = 0
    for w, u, v in edges:
        if dsu.union(u, v):
            mst += w
            used += 1
            if used == n - 1:
                break

    return total - mst

if __name__ == "__main__":
    print(solve("network.txt"))  # 259679 for the official file

File: Java/Problem_107.java
import java.io.*;
import java.util.*;

public class Main {
    static class DSU {
        int[] p, r;
        DSU(int n) {
            p = new int[n];
            r = new int[n];
            for (int i = 0; i < n; i++) p[i] = i;
        }
        int find(int x) {
            while (p[x] != x) {
                p[x] = p[p[x]];
                x = p[x];
            }
            return x;
        }
        boolean union(int a, int b) {
            int ra = find(a), rb = find(b);
            if (ra == rb) return false;
            if (r[ra] < r[rb]) { int t = ra; ra = rb; rb = t; }
            p[rb] = ra;
            if (r[ra] == r[rb]) r[ra]++;
            return true;
        }
    }

    static class Edge implements Comparable<Edge> {
        int w, u, v;
        Edge(int w, int u, int v){ this.w=w; this.u=u; this.v=v; }
        public int compareTo(Edge o){ return Integer.compare(this.w, o.w); }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("network.txt"));
        ArrayList<String[]> rows = new ArrayList<>();
        String line;
        while ((line = br.readLine()) != null) rows.add(line.trim().split(","));
        br.close();

        int n = rows.size();
        ArrayList<Edge> edges = new ArrayList<>();
        long total = 0;

        for (int i = 0; i < n; i++) {
            String[] row = rows.get(i);
            for (int j = i + 1; j < n; j++) {
                String v = row[j];
                if (!v.equals("-")) {
                    int w = Integer.parseInt(v);
                    total += w;
                    edges.add(new Edge(w, i, j));
                }
            }
        }

        Collections.sort(edges);
        DSU dsu = new DSU(n);

        long mst = 0;
        int used = 0;
        for (Edge e : edges) {
            if (dsu.union(e.u, e.v)) {
                mst += e.w;
                used++;
                if (used == n - 1) break;
            }
        }

        System.out.println(total - mst); // 259679 for official file
    }
}
